src/Intervals.jl

```diff
@@ -18,23 +19,52 @@ struct Unbounded <: Bound end
 
 bound_type(x::Bool) = x ? Closed : Open
 
-abstract type AbstractInterval{T, L <: Bound, R <: Bound} end
+# Methods for convert between int and tuple representations for space efficiency
+# (Open, Closed) is 16 bytes while the integer represenation is only 1 byte.
+# TODO: Convert types to symbols when we switch to extending IntervalSets.jl
+bounds_int(l::Type{Open}, r::Type{Open}) = 0x00
+bounds_int(l::Type{Closed}, r::Type{Open}) = 0x01
+bounds_int(l::Type{Open}, r::Type{Closed}) = 0x02
+bounds_int(l::Type{Closed}, r::Type{Closed}) = 0x03
+
+bounds_types(x::Integer) = bounds_types(Val(UInt8(x)))
+bounds_types(::Val{0x00}) = (Open, Open)
+bounds_types(::Val{0x01}) = (Closed, Open)
+bounds_types(::Val{0x02}) = (Open, Closed)
+bounds_types(::Val{0x03}) = (Closed, Closed)
+
+# Extension for backwards support of Unbounded endpoints to avoid changing existing logic
+# TODO: Drop these in favour of the approach in IntervalSets.jl
```

Comment:
Seems like an approach has yet to be worked out: https://github.com/JuliaMath/IntervalSets.jl/issues/123

Reply:
I guess the question is whether it should be the responsibility of either IntervalSets.jl or Intervals.jl to handle this with yet another custom type. AFAICT, bounded-ness is just a question of whether an endpoint value `isfinite` which we already have an API for. My preference would be for that API to be handle independently of intervals like in Infinity.jl or Infinities.jl.

Reply:
I do agree that using `isfinite` on endpoints does seem like a better interface. From what I recall from the Infinity.jl work the parts where this breaks down is that:

1. Defining `Infinite{T}` seems like a great idea for making any type into something that supports infinity but you lose the proper type hierarchy which makes this break down in multiple dispatch. Ideally we'd want to define `Infinite{T} <: supertype(T)` to make this work.
2. With 1 being infeasible we're left with making specific types to properly define supertypes. This results in not all types having an alternative infinite type.

This is where it started to make more sense to have the intervals to support infinity themselves. For example `Interval{Char,Closed,Unbounded}`.

An alternative strategy could be something like:
```julia
struct Interval{T}
    first::Union{T,NegInf}
    last::Union{T,PosInf}
end
```
but I doubt that would be an improvement over the v1 setup.